---
title: "S. aureus sequence analysis"
author: "Leticia Arroyo Benito"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: 2
    toc_float:
      collapse: no
      smooth_scroll: yes
---
 

```{r, include=FALSE}
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(version = "3.19")
BiocManager::install("Biostrings")
library(data.table)
library(ggplot2)
library(Biostrings)
library(ggrepel)
library(ape) #calculo pcoa tb se puede con wcmdscale() de vegan y cmdscale() de stats
library(RColorBrewer)
library(openxlsx)
library(stats)
```

```{r, include=FALSE}
setwd("C:/Leticia/master/TFM")
datos <- "DB_LAB.xlsx"
data<-data.frame(read.xlsx(datos))
data
nt_seqs <- AAStringSet(data$Nucleotide_sequence)
names(nt_seqs) <- data$Name
writeXStringSet(nt_seqs, file="./nt_seqs.fasta")

```


##   AUC TRA TpNa distribution
```{r, echo=FALSE, warning=FALSE}
hist(data$TRA_TpNa_norm, xlab = "AUC TRA_TpNa normalized", main = "")
```

##   AUC TRA HS distribution
```{r, echo=FALSE, warning=FALSE}
hist(data$TRA_HS_norm, xlab = "AUC TRA_HS normalized", main = "")
```

##   AUC Growth inhibition distribution
```{r, echo=FALSE, warning=FALSE}
hist(data$GH_norm, xlab = "AUC Growth inhibition normalized", main = "")

```

##   AUC TRA_TpNa Vs AUC Growth inhibition
```{r, echo=FALSE, warning=FALSE}
ggplot(data) + 
  geom_point(aes(x=as.numeric(GH_norm), y=as.numeric(TRA_TpNa_norm), color = Domain_type), size=3, alpha=0.7) +
  xlab("AUC Growth inhibition") + ylab("AUC TRA TpNa")
```


```{r, include=FALSE}
cor(data$TRA_TpNa_norm, data$GH_norm)

```

## Homology search: MMseqs easy_search

homebrew
leticia@TELUM-PORT2:~$ brew install mmseqs2
leticia@TELUM-PORT2:~$ mmseqs easy-search ./sec.fasta ./sec.fasta ./sec.m8 tmp

Easy-Search
The easy-search workflow enables searching with a FASTA/FASTQ file against another
FASTA/FASTQ file or a pre-built MMseqs2 target database.
mmseqs easy-search examples/QUERY.fasta examples/DB.fasta alnResult.m8 tmp
In this command, 
mmseqs easy-search ./sec.fasta ./sec.fasta ./sec.m8 tmp

sec.fasta is the query file, sec.fasta is the target file, sec.m8 is the file for alignment
results, and tmp is a temporary directory for intermediate files.


```{r, include=FALSE, warning=FALSE}

#importo resultado de mmseqs
ident_df <- fread("./sec.m8")
```
https://www.biostars.org/p/166013/
'qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore'

...
          V1    qseqid means Query Seq-id
          V2    sseqid means Subject Seq-id
          V3    pident means Percentage of identical matches
          V4    length means Alignment length
          V5    mismatch means Number of mismatches
          V6    gapopen means Number of gap openings
          V7    qstart means Start of alignment in query
          V8    qend means End of alignment in query
          V9    sstart means Start of alignment in subject
          V10   send means End of alignment in subject
          V11   evalue means Expect value
          V12   bitscore means Bit score
         
```{r, include=FALSE}
mmseqs2_output_to_pcoa <- function(ident_df, seqs){
  pep_list <- unique(ident_df$V1)
  df <- data.frame(matrix(nrow=length(pep_list), ncol=length(pep_list)))
  colnames(df) <- pep_list
  rownames(df) <- pep_list
  for (k in pep_list){
    for (j in pep_list){
      addr_QS <- which((ident_df$V1 == k) & (ident_df$V2 == j))
      addr_SQ <- which((ident_df$V1 == j) & (ident_df$V2 == k))
      if (isEmpty(addr_QS) & isEmpty(addr_SQ)){
        df[which(rownames(df) == k), which(colnames(df) == j)] <- 1
        df[which(rownames(df) == j), which(colnames(df) == k)] <- 1
      #} else if (isEmpty(addr_QS) & !isEmpty(addr_SQ)){
      #  adj_homology <- ident_df$V3[addr_SQ] * ((ident_df$V8[addr_SQ] - ident_df$V7[addr_SQ] + 1)/seqs@ranges@width[which(seqs@ranges@NAMES==j)])
      #  df[which(rownames(df) == k), which(colnames(df) == j)] <- 1 - adj_homology
      #  df[which(rownames(df) == j), which(colnames(df) == k)] <- 1 - adj_homology
      } else if (!isEmpty(addr_QS) & isEmpty(addr_SQ)){
        adj_homology <- ident_df$V3[addr_QS] * ((ident_df$V8[addr_QS] - ident_df$V7[addr_QS] + 1)/seqs@ranges@width[which(seqs@ranges@NAMES==k)])
        df[which(rownames(df) == k), which(colnames(df) == j)] <- 1 - (adj_homology/2)
        df[which(rownames(df) == j), which(colnames(df) == k)] <- 1 - (adj_homology/2)
      } else if (!isEmpty(addr_QS) & !isEmpty(addr_SQ)){
        adj_homology_QS <- ident_df$V3[addr_QS] * ((ident_df$V8[addr_QS] - ident_df$V7[addr_QS] + 1)/seqs@ranges@width[which(seqs@ranges@NAMES==k)])
        adj_homology_SQ <- ident_df$V3[addr_SQ] * ((ident_df$V8[addr_SQ] - ident_df$V7[addr_SQ] + 1)/seqs@ranges@width[which(seqs@ranges@NAMES==j)])
        df[which(rownames(df) == k), which(colnames(df) == j)] <- 1 - ((adj_homology_QS+adj_homology_SQ)/2)
        df[which(rownames(df) == j), which(colnames(df) == k)] <- 1 - ((adj_homology_QS+adj_homology_SQ)/2)
      }
    }
  }
  dist_df <- as.dist(df)
  ret_list <- pcoa(dist_df)
  return(ret_list)                           
}
```

```{r, include=FALSE}
seqs <- AAStringSet(data$Protein_sequence)
names(seqs) <- data$Name
seqs

```
Each point indicate each peptide, and if two points are closer, it means they are similar in terms of AA sequence.
pcoa() (ape)
```{r, echo=FALSE, warning=FALSE}
pcoa_out <- mmseqs2_output_to_pcoa(ident_df, seqs)
pcoa_axis <- data.frame(pcoa_out$vectors[,c(1,2)])
pcoa_loadings <- pcoa_out$values$Relative_eig[c(1:2)]
pcoa_axis$Domain_type <- as.character(sapply(rownames(pcoa_axis),
                                                 function(x){data$Domain_type[which(data$Name==x)[1]]}))
pcoa_axis$Muralytic_activity <- as.numeric(sapply(rownames(pcoa_axis),
                                                     function(x){data$TRA_TpNa_norm[which(data$Name==x)[1]]}))
pcoa_axis$Growth <- sapply(rownames(pcoa_axis),
                           function(x){data$GH_norm[which(data$Name==x)[1]]})
pcoa_axis$Domain_name <- rownames(pcoa_axis)


ggplot(pcoa_axis) +
  geom_point(aes(x=Axis.1, y=Axis.2, color = Domain_type), size=3, alpha=0.7) + 
  xlab(paste0("Axis 1 (", round(pcoa_loadings[1]*100, 2), ")")) +
  ylab(paste0("Axis 2 (", round(pcoa_loadings[2]*100, 2), ")"))


#pcoa_out$values$...
#Eigenvalues: All eigenvalues (pstve, null and ngatve)
#Relative_eig: relatve eienvalues
#Rel_corr_eig: Relative eigenvalues after Lingoes or Cailliez correction (correcion xa no tener eigenv ngatvos)
#Broken_stick: Expected fractions of variance under the broken stick model.
#Cum_corr_eig: Cumulative corrected relative eigenvalues
#Cumul_br_stick: Cumulative broken stick fractions
```


## Normalized AUC TRA_TpNa gradient

Labelled those proteins whose normalized AUC is <0
```{r, echo=FALSE, warning=FALSE}
ggplot(pcoa_axis[order(as.numeric(pcoa_axis$Muralytic_activity), decreasing=F),]) + 
  geom_point(aes(x=Axis.1, y=Axis.2, color= Muralytic_activity), size=2, alpha=0.6) +
  geom_text_repel(data=pcoa_axis[which(pcoa_axis$Muralytic_activity<0),],
                  aes(x=Axis.1, y=Axis.2, label= paste(Domain_name, ":", Domain_type)), 
                  nudge_x=c(-0.2, 0.2), nudge_y=c(-0.2, 0.2), direction="both", size=2) +
  scale_color_gradient(low="blue", high="red")+
  theme(legend.title = element_text(size = 10, hjust = 0.5)
  )+ labs(color = "AUC TRA_TpNa \nnormalized")

```


##  Normalized Growth inhibition gradient

Labelled those proteins whose normalized AUC is <0
```{r, echo=FALSE, warning=FALSE}
ggplot(pcoa_axis[order(as.numeric(pcoa_axis$Growth), decreasing=F),]) + 
  geom_point(aes(x=Axis.1, y=Axis.2, color=Growth), size=2, alpha=0.6) +
  geom_text_repel(data=pcoa_axis[which(pcoa_axis$Growth<0),],
                  aes(x=Axis.1, y=Axis.2, label= paste(Domain_name, ":", Domain_type)), 
                  nudge_x=c(-0.2, 0.2), nudge_y=c(-0.2, 0.2), direction="both", size=2) +
  scale_color_gradient(low="blue", high="red")+
  theme(legend.title = element_text(size = 10, hjust = 0.5)
  )+ labs(color = "AUC Growth inhibition \nnormalized")

```

##  MSA geneious 
StatQuest: MDS and PCoA in R: youtube.com/watch?v=pGAUHhLYp5Q
cmdscale
metric to calculate distance matrix: Euclidean distance
```{r, include=FALSE}
msa <- read.csv("msa.csv", sep = ";")
row.names(msa) <- msa$Prote
msa$Prote <- NULL
#create distance matrix
#directly geneious identity matrix
msa_matrix <- as.matrix(msa)
msa_dist <- dist(msa_matrix)
msa_dist

#multidimensional scaling, 2 primeros. add=TRUE xa eigenvalues pstve
cmd_msa <- cmdscale(msa_dist, eig = TRUE, k=2, add = TRUE)
print(cmd_msa)
positions <- cmd_msa$points
# amount of variation explained by the different axes. Using eigenvalues
var_explained <- round(100*cmd_msa$eig/sum(cmd_msa$eig), 1)
var_explained

#format the data for ggplot

mds.data <- data.frame(Sample=rownames(positions), X=positions[,1], Y=positions[,2])
mds.data

#merge with mmseqs df 
colnames(mds.data)[1] <- "Domain_name"
df3 <- merge(x = mds.data, y = pcoa_axis, by="Domain_name")
df3

```

```{r, echo=FALSE, warning=FALSE}

ggplot(data = df3, aes(x=X, y=Y, color=Domain_type))+
  geom_point()+ theme_bw()+ #geom_text()+
  xlab(paste("MDS1 (", var_explained[1], "%)", sep = ""))+
  ylab(paste("MDS2 (", var_explained[2], "%)", sep = ""))
```
